packages/headless/src/controllers/generated-answer/headless-generated-answer.ts

```diff
@@ -0,0 +1,132 @@
+import {EventSourcePolyfill} from 'event-source-polyfill';
+import {GeneratedAnswerCitation} from '../../api/generated-answer/generated-answer-event-payload';
+import {SearchEngine} from '../../app/search-engine/search-engine';
+import {
+  streamAnswer,
+  resetAnswer,
+} from '../../features/generated-answer/generated-answer-actions';
+import {
+  logDislikeGeneratedAnswer,
+  logLikeGeneratedAnswer,
+  logOpenGeneratedAnswerSource,
+  logRetryGeneratedAnswer,
+} from '../../features/generated-answer/generated-answer-analytics-actions';
+import {generatedAnswerReducer as generatedAnswer} from '../../features/generated-answer/generated-answer-slice';
+import {GeneratedAnswerState} from '../../features/generated-answer/generated-answer-state';
+import {executeSearch} from '../../features/search/search-actions';
+import {GeneratedAnswerSection} from '../../state/state-sections';
+import {loadReducerError} from '../../utils/errors';
+import {buildController} from '../controller/headless-controller';
+
+export type {GeneratedAnswerState, GeneratedAnswerCitation};
+
+/**
+ * @internal
+ */
+export interface GeneratedAnswer {
+  /**
+   * The state of the GeneratedAnswer controller.
+   */
+  state: GeneratedAnswerState;
+  /**
+   * Re-executes the last query to generate an answer.
+   */
+  retry(): void;
+  /**
+   * Logs a custom event indicating a generated answer is relevant.
+   */
+  likeGeneratedAnswer(): void;
+  /**
+   * Logs a custom event indicating a generated answer is irrelevant.
+   */
+  dislikeGeneratedAnswer(): void;
+  /**
+   * Logs a custom event indicating a cited source link was clicked.
+   */
+  logCitationClick(citation: GeneratedAnswerCitation): void;
+}
+
+/**
+ * @internal
+ */
+export function buildGeneratedAnswer(engine: SearchEngine): GeneratedAnswer {
+  if (!loadGeneratedAnswerReducer(engine)) {
+    throw loadReducerError;
+  }
+
+  const {dispatch} = engine;
+  const controller = buildController(engine);
+  const getState = () => engine.state;
+
+  let source: EventSourcePolyfill;
+  let lastRequestId: string;
+  let lastStreamId: string;
+
+  const setEventSourceRef = (sourceRef: EventSourcePolyfill) => {
+    source = sourceRef;
+  };
+
+  const getIsStreamInProgress = () =>
+    source &&
+    (source.readyState === source.OPEN ||
+      source.readyState === source.CONNECTING);
+
+  const subscribeToSearchRequests = () => {
+    const strictListener = () => {
+      const state = getState();
+      const requestId = state.search.requestId;
+      const streamId =
+        state.search.extendedResults.generativeQuestionAnsweringId;
```

Comment:
I vote to drop the "Question" part of the name to make it shorter :)

Reply:
Actually just noticed too... but the SmartSnippet feature is actually called `Question Answering` in the code, so to avoid confusion, I would definitely stick with `Generated Answer`

Reply:
I agree and it's `generatedAnswer` everywhere except in the param the search api returns us that contains the stream ID. That one is called `generativeQuestionAnsweringId` and it's the only place.